src/compute-engine/boxed-expression/match.ts

```diff
@@ -352,82 +403,217 @@ function matchArguments(
+  return matchRemaining(patterns, substitution);
+
+  /*
+   * Local f().
+   */
+  /**
+   * Match a list of patterns against *remaining* (locally scoped) `ops` and consume this list along
+   * the way, whilst appending to `substitution`. If a complete/successful match, return this new
+   * substitution; else return `null`.
+   *
+   * @note: calls recursively (permutations) for sequence wildcards
+   *
+   * @param patterns
+   * @param substitution
+   * @returns
+   */
+  function matchRemaining(
+    patterns: ReadonlyArray<BoxedExpression>,
+    substitution: BoxedSubstitution
+  ): BoxedSubstitution | null {
+    let result: BoxedSubstitution | null = { ...substitution };
+
+    let i = 0; // Index in pattern
+
+    while (i < patterns.length) {
+      const pat = patterns[i];
+      const argName = wildcardName(pat);
+
+      if (argName !== null) {
+        if (argName.startsWith('__')) {
+          // Match 1 or more expressions (__) or 0 or more (___)
+          const nextPattern = patterns[i + 1];
+
+          const isOptionalSeq = argName.startsWith('___');
+
+          if (nextPattern === undefined) {
+            // No more args in the pattern after, go till the end
+            if (ops.length === 0 && !isOptionalSeq) return null;
+            result = captureWildcard(argName, captureOps(ops.length), result);
           } else {
-            value = ce.function('Sequence', args, { canonical: false });
+            /** Total qty. of operands to be consumed by this sequence (of those remaining).
+             * If a non-optional sequence, minimum must be 1.
+             */
+            let j = isOptionalSeq ? 0 : 1;
+
+            // The next pattern should not be another required sequence wildcard
+            // (^@todo?: validate beforehand; should never be permitted?)
+            console.assert(
+              !(
+                isWildcard(nextPattern) &&
+                wildcardType(nextPattern) === 'Sequence'
+              )
+            );
+
+            // The next 'applicable' pattern.
+            // If this sequence is qualified by one, or more optional-sequence wildcards, then these
+            // should effectively be 'merged' with this sequence, and hence be marked as capturing
+            // '0' operands. In this case, skip over these, arriving at the next applicable pattern,
+            // and capture the optional wildcards subsequently.
+            let nextAppPattern = nextPattern;
+            let nextAppPatternIndex = i + 1;
+
+            while (
+              isWildcard(nextAppPattern) &&
+              wildcardType(nextAppPattern) === 'OptionalSequence'
+            ) {
+              if (!patterns[nextAppPatternIndex + 1]) break;
+              // @note: if pattern has been validated prior, the next should never be a '_'
+              // (universal) or '__' (regular sequence) wildcard
+              nextAppPattern = patterns[++nextAppPatternIndex];
+            }
```

Comment:
This is a relatively big added change/check, but believe procedure here is sensible, if you want to verify... ?